src/event_parser.py

```diff
@@ -0,0 +1,205 @@
+import base64
+from hashlib import sha256
+from typing import Any, Dict, Union, Callable
+from dataclasses import dataclass
+import struct
+
+bytes_map = {
+    'bool': 1,
+    'u8': 1,
+    'i8': 1,
+    'u16': 2,
+    'i16': 2,
+    'u32': 4,
+    'i32': 4,
+    'f32': 4,
+    'u64': 8,
+    'i64': 8,
+    'f64': 8,
+    'u128': 16,
+    'i128': 16,
+    'u256': 32,
+    'i256': 32,
+    'bytes': None,
+    'string': None,
+    'publicKey': 32,
+    'enum': 1
+}
+
+def convert_bytes_to_type(data_type):
+    conversion_functions = {
+        'bool': lambda b: bool(int.from_bytes(b, 'little')),
+        'u8': lambda b: b[0],
+        'i8': lambda b: int.from_bytes(b, 'little', signed=True),
+        'u16': lambda b: int.from_bytes(b, 'little'),
+        'i16': lambda b: int.from_bytes(b, 'little', signed=True),
+        'u32': lambda b: int.from_bytes(b, 'little'),
+        'i32': lambda b: int.from_bytes(b, 'little', signed=True),
+        'f32': lambda b: struct.unpack('<f', b)[0],
+        'u64': lambda b: int.from_bytes(b, 'little'),
+        'i64': lambda b: int.from_bytes(b, 'little', signed=True),
+        'f64': lambda b: struct.unpack('<d', b)[0],
+        'u128': lambda b: int.from_bytes(b, 'little'),
+        'i128': lambda b: int.from_bytes(b, 'little', signed=True),
+        'u256': lambda b: int.from_bytes(b, 'little'),
+        'i256': lambda b: int.from_bytes(b, 'little', signed=True),
+        'bytes': lambda b: b,
+        'string': lambda b: b.decode('utf-8'),
+        'publicKey': lambda b: b,
+        'enum': lambda b: int.from_bytes(b, 'little')  # TODO get variants as input
+    }
+    return conversion_functions[data_type]
+
+
+def camelcase(s):
+    """
+    Convert a string into camelCase.
+    """
+    parts = s.split()
+    # Lowercase the first word and capitalize subsequent words
+    return ''.join([parts[0].lower()] + [word.capitalize() for word in parts[1:]])
+
+
+@dataclass
+class Layout:
+    name: str
+    conversion_function: Callable
+    length: int
+
+    def __str__(self):
+        return f"Layout: {self.name} \n   Conversion function: {self.conversion_function} \n   Bytes: {self.length}"
+
+    def decode(self):
+        raise NotImplementedError
+
+
+class EventDecoder:
+    def __init__(self, idl: Dict):
+        self.layouts: Dict[str, Layout] = dict()
```

Comment:
Some of the events seem to have a list of layouts which makes the `decode` method fail on `AttributeError`.